File: scout/scoring.py
"""
AutoMedal — Competition Heuristic Scorer
==========================================
Two-stage scoring system for ranking Kaggle competitions by AutoMedal compatibility.

Stage 1: Metadata-only scoring (from competitions_list() API response)
Stage 2: File listing check (for top-N candidates, lightweight API call)

No LLM calls. No data downloads. Pure deterministic heuristics.
"""

import re
from datetime import datetime, timezone

# ─── KNOWN TABULAR METRICS ─────────────────────────────────────────────
# Case-insensitive matching. Kaggle API uses various formats.
KNOWN_TABULAR_METRICS = {
    "accuracy", "logloss", "log_loss", "categoricalcrossentropy",
    "categoricalaccuracy", "auc", "rmse", "mae", "r2", "f1",
    "rootmeansquarederror", "rootmeansquaredlogerror",
    "meanabsoluteerror", "meansquarederror", "mse",
    "multiclassloss", "mapk", "map@k", "quadraticweightedkappa",
    "weightedlogloss", "binarycrossentropy", "f1score",
    "rmsle", "medianabsoluteerror", "spearman",
}

# ─── NON-TABULAR TAG KEYWORDS ──────────────────────────────────────────
NON_TABULAR_TAGS = {
    "nlp", "computer vision", "image classification", "object detection",
    "text", "audio", "video", "segmentation", "image",
    "natural language processing", "speech", "generative ai",
    "large language models", "diffusion", "gan",
}

# ─── NON-TABULAR DESCRIPTION KEYWORDS ──────────────────────────────────
NON_TABULAR_DESC_KEYWORDS = [
    r"\bimage\b", r"\bnlp\b", r"\btext classification\b",
    r"\bsegmentation\b", r"\bobject detection\b", r"\bcomputer vision\b",
    r"\bspeech\b", r"\baudio\b", r"\bvideo\b", r"\bpixel\b",
    r"\btoken\b", r"\btransformer\b", r"\bbert\b", r"\bgpt\b",
    r"\bllm\b", r"\bdiffusion\b",
]

# ─── NON-TABULAR FILE EXTENSIONS ───────────────────────────────────────
NON_TABULAR_EXTENSIONS = {
    ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".tiff", ".webp",
    ".mp3", ".wav", ".flac", ".mp4", ".avi", ".mov",
    ".json", ".jsonl", ".txt", ".parquet",
}


def _normalize_metric(metric):
    """Normalize a metric string for comparison."""
    if not metric:
        return ""
    return re.sub(r"[^a-z0-9]", "", metric.lower())


def _days_until_deadline(deadline_str):
    """Calculate days until deadline. Returns None if unparseable."""
    if not deadline_str:
        return None
    try:
        # Kaggle API returns ISO format or similar
        if isinstance(deadline_str, datetime):
            deadline = deadline_str
        else:
            deadline_str = str(deadline_str).replace("Z", "+00:00")
            deadline = datetime.fromisoformat(deadline_str)
        if deadline.tzinfo is None:
            deadline = deadline.replace(tzinfo=timezone.utc)
        now = datetime.now(timezone.utc)
        return (deadline - now).days
    except (ValueError, TypeError):
        return None


def _get_tags_lower(competition):
    """Extract lowercase tag names from competition metadata."""
    tags = competition.get("tags", [])
    if not tags:
        return set()
    result = set()
    for tag in tags:
        if isinstance(tag, str):
            result.add(tag.lower())
        elif isinstance(tag, dict):
            name = tag.get("name", "") or tag.get("ref", "")
            if name:
                result.add(name.lower())
        elif hasattr(tag, "name"):
            result.add(tag.name.lower())
        elif hasattr(tag, "ref"):
            result.add(tag.ref.lower())
    return result


# ─── STAGE 1: METADATA-ONLY SCORING ────────────────────────────────────

def score_stage1(competition):
    """Score a competition using only API metadata (no downloads).

    Args:
        competition: dict-like object from Kaggle API competitions_list().

    Returns:
        (score: int, reasons: list[str], disqualified: bool)
    """
    score = 0
    reasons = []

    # --- Extract fields (handle both dict and object access) ---
    def _get(key, default=None):
        if isinstance(competition, dict):
            return competition.get(key, default)
        return getattr(competition, key, default)

    ref = str(_get("ref", "") or "")
    title = str(_get("title", "") or "")
    description = str(_get("description", "") or "")
    category = str(_get("category", "") or "")
    eval_metric = str(_get("evaluationMetric", "") or "")
    team_count = _get("teamCount", 0) or 0
    kernels_only = _get("isKernelsSubmissionsOnly", False)
    deadline = _get("deadline", None)

    tags_lower = _get_tags_lower(competition)

    # --- HARD DISQUALIFIERS ---
    if kernels_only:
        return (0, ["DISQUALIFIED: kernels-only submissions"], True)

    days_left = _days_until_deadline(deadline)
    if days_left is not None and days_left < 7:
        return (0, [f"DISQUALIFIED: deadline in {days_left} days"], True)

    # --- POSITIVE SIGNALS ---

    # Playground Series slug pattern (+40)
    if "playground-series" in ref.lower():
        score += 40
        reasons.append("+40: Playground Series (slug match)")

    # Category bonuses
    cat_lower = category.lower()
    if cat_lower == "getting started":
        score += 30
        reasons.append("+30: Getting Started category")
    elif cat_lower == "playground":
        score += 25
        reasons.append("+25: Playground category")
    elif cat_lower == "featured":
        score += 5
        reasons.append("+5: Featured category")

    # Known tabular evaluation metric (+20)
    metric_normalized = _normalize_metric(eval_metric)
    if metric_normalized and metric_normalized in {
        _normalize_metric(m) for m in KNOWN_TABULAR_METRICS
    }:
        score += 20
        reasons.append(f"+20: known tabular metric ({eval_metric})")

    # Tags contain "tabular" (+20)
    if "tabular" in tags_lower:
        score += 20
        reasons.append("+20: tagged 'tabular'")

    # Team count signals
    if team_count > 2000:
        score += 15
        reasons.append(f"+15: very active ({team_count} teams)")
    elif team_count > 500:
        score += 10
        reasons.append(f"+10: active ({team_count} teams)")
    elif team_count > 100:
        score += 5
        reasons.append(f"+5: moderate activity ({team_count} teams)")

    # Deadline health
    if days_left is not None and days_left > 30:
        score += 5
        reasons.append(f"+5: comfortable deadline ({days_left} days)")

    # --- NEGATIVE SIGNALS ---

    # Non-tabular tags (-50)
    bad_tags = tags_lower & NON_TABULAR_TAGS
    if bad_tags:
        score -= 50
        reasons.append(f"-50: non-tabular tags ({', '.join(bad_tags)})")

    # Non-tabular description keywords (-30)
    desc_lower = description.lower()
    matched_keywords = [
        kw for kw in NON_TABULAR_DESC_KEYWORDS
        if re.search(kw, desc_lower)
    ]
    if matched_keywords and not bad_tags:
        # Only apply if tags didn't already catch it
        score -= 30
        reasons.append(f"-30: non-tabular description keywords")

    # Unknown/missing evaluation metric (-10)
    if not eval_metric:
        score -= 10
        reasons.append("-10: no evaluation metric specified")
    elif metric_normalized and metric_normalized not in {
        _normalize_metric(m) for m in KNOWN_TABULAR_METRICS
    }:
        score -= 5
        reasons.append(f"-5: unknown metric ({eval_metric})")

    # Research category (-20)
    if cat_lower == "research":
        score -= 20
        reasons.append("-20: Research category (non-standard)")

    return (score, reasons, False)


# ─── STAGE 2: FILE LISTING CHECK ───────────────────────────────────────

def score_stage2(file_list):
    """Score a competition based on its file listing (no data download).

    Args:
        file_list: list of dicts with 'name' and optionally 'totalBytes' keys.
                   Or list of objects with .name and .totalBytes attributes.

    Returns:
        (score: int, reasons: list[str])
    """
    score = 0
    reasons = []

    if not file_list:
        reasons.append("+0: no file listing available")
        return (score, reasons)

    # Extract file info
    filenames = []
    total_bytes = 0
    for f in file_list:
        if isinstance(f, dict):
            name = f.get("name", "")
            size = f.get("totalBytes", 0) or 0
        else:
            name = getattr(f, "name", "")
            size = getattr(f, "totalBytes", 0) or 0
        filenames.append(name.lower())
        total_bytes += size

    extensions = {
        "." + fn.rsplit(".", 1)[-1] if "." in fn else ""
        for fn in filenames
    }

    # train.csv present (+25)
    has_train = any(
        fn in ("train.csv", "train.csv.zip", "train.csv.gz")
        for fn in filenames
    )
    if has_train:
        score += 25
        reasons.append("+25: train.csv found")

    # test.csv present (+15)
    has_test = any(
        fn in ("test.csv", "test.csv.zip", "test.csv.gz")
        for fn in filenames
    )
    if has_test:
        score += 15
        reasons.append("+15: test.csv found")

    # sample_submission.csv present (+10)
    has_sample = any(
        "sample" in fn and "submission" in fn
        for fn in filenames
    )
    if has_sample:
        score += 10
        reasons.append("+10: sample_submission.csv found")

    # All CSV files (+10)
    csv_exts = {".csv", ".csv.zip", ".csv.gz", ".zip"}
    non_csv = extensions - csv_exts - {""}
    if not non_csv and filenames:
        score += 10
        reasons.append("+10: all files are CSV/zip")

    # Non-tabular file types (-50)
    bad_exts = extensions & NON_TABULAR_EXTENSIONS
    if bad_exts:
        score -= 50
        reasons.append(f"-50: non-tabular files ({', '.join(bad_exts)})")

    # Total size > 5GB (-15)
    size_gb = total_bytes / (1024 ** 3)
    if size_gb > 5:
        score -= 15
        reasons.append(f"-15: large dataset ({size_gb:.1f}GB)")

    return (score, reasons)


# ─── COMBINED SCORING ───────────────────────────────────────────────────

def compute_final_score(stage1_score, stage2_score):
    """Clamp combined score to [0, 100]."""
    return max(0, min(100, stage1_score + stage2_score))

File: internal/scout/scoring.go
// Two-stage heuristic scoring. Mirrors scout/scoring.py.
package scout

import (
	"regexp"
	"strings"
	"time"
)

// Competition is the metadata shape used by scoring + select. Mirrors the
// dict produced by scout/discover.py:_competition_to_dict.
type Competition struct {
	Ref                      string     `json:"ref"`
	Title                    string     `json:"title"`
	Description              string     `json:"description"`
	URL                      string     `json:"url"`
	Category                 string     `json:"category"`
	Reward                   any        `json:"reward,omitempty"`
	Deadline                 string     `json:"deadline,omitempty"`
	TeamCount                int        `json:"teamCount"`
	IsKernelsSubmissionsOnly bool       `json:"isKernelsSubmissionsOnly"`
	EvaluationMetric         string     `json:"evaluationMetric"`
	MaxDailySubmissions      any        `json:"maxDailySubmissions,omitempty"`
	MaxTeamSize              any        `json:"maxTeamSize,omitempty"`
	EnabledDate              string     `json:"enabledDate,omitempty"`
	Tags                     []string   `json:"tags"`
	Files                    []FileInfo `json:"files,omitempty"`
}

// FileInfo describes one competition file.
type FileInfo struct {
	Name       string `json:"name"`
	TotalBytes int64  `json:"totalBytes"`
}

var (
	knownTabularMetrics = []string{
		"accuracy", "logloss", "log_loss", "categoricalcrossentropy",
		"categoricalaccuracy", "auc", "rmse", "mae", "r2", "f1",
		"rootmeansquarederror", "rootmeansquaredlogerror",
		"meanabsoluteerror", "meansquarederror", "mse",
		"multiclassloss", "mapk", "map@k", "quadraticweightedkappa",
		"weightedlogloss", "binarycrossentropy", "f1score",
		"rmsle", "medianabsoluteerror", "spearman",
	}
	nonTabularTags = map[string]bool{
		"nlp": true, "computer vision": true, "image classification": true,
		"object detection": true, "text": true, "audio": true, "video": true,
		"segmentation": true, "image": true,
		"natural language processing": true, "speech": true,
		"generative ai": true, "large language models": true,
		"diffusion": true, "gan": true,
	}
	nonTabularDescPatterns = []*regexp.Regexp{
		regexp.MustCompile(`\bimage\b`), regexp.MustCompile(`\bnlp\b`),
		regexp.MustCompile(`\btext classification\b`), regexp.MustCompile(`\bsegmentation\b`),
		regexp.MustCompile(`\bobject detection\b`), regexp.MustCompile(`\bcomputer vision\b`),
		regexp.MustCompile(`\bspeech\b`), regexp.MustCompile(`\baudio\b`),
		regexp.MustCompile(`\bvideo\b`), regexp.MustCompile(`\bpixel\b`),
		regexp.MustCompile(`\btoken\b`), regexp.MustCompile(`\btransformer\b`),
		regexp.MustCompile(`\bbert\b`), regexp.MustCompile(`\bgpt\b`),
		regexp.MustCompile(`\bllm\b`), regexp.MustCompile(`\bdiffusion\b`),
	}
	nonTabularExtensions = map[string]bool{
		".jpg": true, ".jpeg": true, ".png": true, ".gif": true, ".bmp": true,
		".tiff": true, ".webp": true, ".mp3": true, ".wav": true, ".flac": true,
		".mp4": true, ".avi": true, ".mov": true,
		".json": true, ".jsonl": true, ".txt": true, ".parquet": true,
	}
	nonAlnumRE = regexp.MustCompile(`[^a-z0-9]`)
)

func normalizeMetric(s string) string {
	if s == "" {
		return ""
	}
	return nonAlnumRE.ReplaceAllString(strings.ToLower(s), "")
}

// daysUntilDeadline returns the day count (negative when past). -1 means
// unparseable; the caller treats -1 like nil in the Python impl.
func daysUntilDeadline(s string) (days int, ok bool) {
	if s == "" {
		return 0, false
	}
	cleaned := strings.ReplaceAll(s, "Z", "+00:00")
	t, err := time.Parse(time.RFC3339, cleaned)
	if err != nil {
		t, err = time.Parse("2006-01-02T15:04:05.999999-07:00", cleaned)
		if err != nil {
			return 0, false
		}
	}
	return int(time.Until(t.UTC()).Hours() / 24), true
}

func tagsLower(tags []string) map[string]bool {
	out := map[string]bool{}
	for _, t := range tags {
		if t = strings.ToLower(strings.TrimSpace(t)); t != "" {
			out[t] = true
		}
	}
	return out
}

// ScoreStage1 scores a competition by metadata only. Returns (score,
// reasons, disqualified).
func ScoreStage1(c Competition) (int, []string, bool) {
	score := 0
	var reasons []string

	if c.IsKernelsSubmissionsOnly {
		return 0, []string{"DISQUALIFIED: kernels-only submissions"}, true
	}
	if days, ok := daysUntilDeadline(c.Deadline); ok && days < 7 {
		return 0, []string{stringf("DISQUALIFIED: deadline in %d days", days)}, true
	}

	if strings.Contains(strings.ToLower(c.Ref), "playground-series") {
		score += 40
		reasons = append(reasons, "+40: Playground Series (slug match)")
	}
	cat := strings.ToLower(c.Category)
	switch cat {
	case "getting started":
		score += 30
		reasons = append(reasons, "+30: Getting Started category")
	case "playground":
		score += 25
		reasons = append(reasons, "+25: Playground category")
	case "featured":
		score += 5
		reasons = append(reasons, "+5: Featured category")
	}

	metricNorm := normalizeMetric(c.EvaluationMetric)
	knownNorms := map[string]bool{}
	for _, m := range knownTabularMetrics {
		knownNorms[normalizeMetric(m)] = true
	}
	if metricNorm != "" && knownNorms[metricNorm] {
		score += 20
		reasons = append(reasons, stringf("+20: known tabular metric (%s)", c.EvaluationMetric))
	}

	tl := tagsLower(c.Tags)
	if tl["tabular"] {
		score += 20
		reasons = append(reasons, "+20: tagged 'tabular'")
	}

	switch {
	case c.TeamCount > 2000:
		score += 15
		reasons = append(reasons, stringf("+15: very active (%d teams)", c.TeamCount))
	case c.TeamCount > 500:
		score += 10
		reasons = append(reasons, stringf("+10: active (%d teams)", c.TeamCount))
	case c.TeamCount > 100:
		score += 5
		reasons = append(reasons, stringf("+5: moderate activity (%d teams)", c.TeamCount))
	}

	if days, ok := daysUntilDeadline(c.Deadline); ok && days > 30 {
		score += 5
		reasons = append(reasons, stringf("+5: comfortable deadline (%d days)", days))
	}

	badTags := []string{}
	for t := range tl {
		if nonTabularTags[t] {
			badTags = append(badTags, t)
		}
	}
	if len(badTags) > 0 {
		score -= 50
		reasons = append(reasons, stringf("-50: non-tabular tags (%s)", strings.Join(badTags, ", ")))
	}

	descLower := strings.ToLower(c.Description)
	hasDescMatch := false
	for _, re := range nonTabularDescPatterns {
		if re.FindStringIndex(descLower) != nil {
			hasDescMatch = true
			break
		}
	}
	if hasDescMatch && len(badTags) == 0 {
		score -= 30
		reasons = append(reasons, "-30: non-tabular description keywords")
	}

	if c.EvaluationMetric == "" {
		score -= 10
		reasons = append(reasons, "-10: no evaluation metric specified")
	} else if metricNorm != "" && !knownNorms[metricNorm] {
		score -= 5
		reasons = append(reasons, stringf("-5: unknown metric (%s)", c.EvaluationMetric))
	}

	if cat == "research" {
		score -= 20
		reasons = append(reasons, "-20: Research category (non-standard)")
	}

	return score, reasons, false
}

// ScoreStage2 scores a competition by file listing. (score, reasons).
func ScoreStage2(files []FileInfo) (int, []string) {
	if len(files) == 0 {
		return 0, []string{"+0: no file listing available"}
	}
	score := 0
	var reasons []string

	totalBytes := int64(0)
	exts := map[string]bool{}
	names := []string{}
	for _, f := range files {
		n := strings.ToLower(f.Name)
		names = append(names, n)
		totalBytes += f.TotalBytes
		if i := strings.LastIndex(n, "."); i >= 0 {
			exts["."+n[i+1:]] = true
		}
	}

	hasName := func(targets ...string) bool {
		for _, n := range names {
			for _, t := range targets {
				if n == t {
					return true
				}
			}
		}
		return false
	}

	if hasName("train.csv", "train.csv.zip", "train.csv.gz") {
		score += 25
		reasons = append(reasons, "+25: train.csv found")
	}
	if hasName("test.csv", "test.csv.zip", "test.csv.gz") {
		score += 15
		reasons = append(reasons, "+15: test.csv found")
	}
	hasSample := false
	for _, n := range names {
		if strings.Contains(n, "sample") && strings.Contains(n, "submission") {
			hasSample = true
			break
		}
	}
	if hasSample {
		score += 10
		reasons = append(reasons, "+10: sample_submission.csv found")
	}

	csvOK := map[string]bool{".csv": true, ".csv.zip": true, ".csv.gz": true, ".zip": true}
	allCSV := len(names) > 0
	for e := range exts {
		if e == "" {
			continue
		}
		if !csvOK[e] {
			allCSV = false
			break
		}
	}
	if allCSV {
		score += 10
		reasons = append(reasons, "+10: all files are CSV/zip")
	}

	bad := []string{}
	for e := range exts {
		if nonTabularExtensions[e] {
			bad = append(bad, e)
		}
	}
	if len(bad) > 0 {
		score -= 50
		reasons = append(reasons, stringf("-50: non-tabular files (%s)", strings.Join(bad, ", ")))
	}

	sizeGB := float64(totalBytes) / (1024 * 1024 * 1024)
	if sizeGB > 5 {
		score -= 15
		reasons = append(reasons, stringf("-15: large dataset (%.1fGB)", sizeGB))
	}
	return score, reasons
}

// ComputeFinalScore clamps to [0, 100].
func ComputeFinalScore(s1, s2 int) int {
	v := s1 + s2
	if v < 0 {
		return 0
	}
	if v > 100 {
		return 100
	}
	return v
}
